chapter_computer-vision/image-augmentation.md

```diff
@@ -359,12 +359,17 @@ def train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs,
 #@tab pytorch
 #@save
 def train_ch13(net, train_iter, test_iter, loss, trainer, num_epochs,
-               devices=d2l.try_all_gpus()):
-    """Train a model with mutiple GPUs (defined in Chapter 13)."""
+               devices=d2l.try_all_gpus(), lazy=False):
+    """Train a model with mutiple GPUs (defined in Chapter 13).
+
+    Defined in :numref:`sec_image_augmentation`"""
     timer, num_batches = d2l.Timer(), len(train_iter)
     animator = d2l.Animator(xlabel='epoch', xlim=[1, num_epochs], ylim=[0, 1],
                             legend=['train loss', 'train acc', 'test acc'])
-    net = nn.DataParallel(net, device_ids=devices).to(devices[0])
+    if lazy:
+        net = net.to(devices[0])
```

Comment:
We should still support multi-GPU training when using Lazy APIs if possible. 
@AnirudhDagar Do you have ideas on this? Besides, did you remember why we have `nn.DataParallel(net, device_ids=devices).to(devices[0])`? Since we have `net = nn.DataParallel(net, device_ids=devices)` in http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_computational-performance/multiple-gpus-concise.html ,  is there any special consideration for using `to(devices[0])`?

Reply:
Q1.
To answer your first question we need to understand how `nn.DataParallel` works fundamentally.

When using `nn.DataParallel`, data should be initially stored on one GPU (the “primary GPU”). This also calls for the model to be moved to the GPU as the parameters need to be on the device for the forward pass. If there are no parameters initialized in the Pytorch layer (Lazy implementation) then it is technically not possible to copy those parameters when the dataparallel runs, essentially its four steps internally: `replicate`, `scatter`, `gather`, `parallel_apply`. Hence I don't think there is a direct way to use `nn.DataParallel` with Lazy implementation without a dry run of input dummy data into the model.

As mentioned in the docs, the expected way to use lazy API is to have a dry run, but in our case, this might lead to slightly not so good looking code:
> After construction, networks with lazy modules should first be converted to the desired dtype and placed on the expected device. This is because lazy modules only perform shape inference so the usual dtype and device placement behavior applies. The lazy modules should then perform “dry runs” to initialize all the components in the module. These “dry runs” send inputs of the correct size, dtype, and device through the network and to each one of its lazy modules. After this the network can be used as usual.

Q2.

> Besides, did you remember why we have nn.DataParallel(net, device_ids=devices).to(devices[0])? Since we have net = nn.DataParallel(net, device_ids=devices) in http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_computational-performance/multiple-gpus-concise.html , is there any special consideration for using to(devices[0])?

This is a very good question, in the [multiple gpus concise chapter](https://d2l.ai/chapter_computational-performance/multiple-gpus-concise.html#summary) we were able to use `net = nn.DataParallel(net, device_ids=devices)` without `net = nn.DataParallel(net, device_ids=devices).to(devices[0])`  because after the first training run (running with single GPU) the network params are already on GPU. When passing a single GPU dataparallel can move the model to that device. And then we don't need to explicitly move them to device[0] when we call training (run with two GPUs).

To be more precise if you skip the first training call, then the second function in that notebook should fail complaining about module parameters not residing on device[0]. This can be fixed if we also change `net = nn.DataParallel(net, device_ids=devices)` to `net = nn.DataParallel(net, device_ids=devices).to(devices[0])`

Q3.
Why `device[0]`?
There is no special reason, if we were to call only the dataparallel with 2 gpus (i.e the second train call with num_gpus=2) then we can move the parameters to any device, for that matter, `nn.DataParallel(net, device_ids=devices).to(devices[1])` should also work as long as the input tensors are also moved to Device 1.
But, we use devices[0] to make sure both functions work (because when calling train with `num_gpus=1` since there will be only one device 0 available), as a simple fix we write code in a fashion that is compatible to the scenario when only one device is available i.e (the first train call with num_gpus=1).


I hope this explanation clears the questions.

Reply:
Thanks for the detailed explanations!

Reply:
Thanks for the explanations! I still have a question here. Will the training process run on a single GPU or multiple GPUs after we explicitly put dataparallel to devices[0]?

Reply:
It will run on multiple GPUs if we pass multiple GPUs to the argument `device_ids `. Think of it as moving the network to the GPU device 0 and since dataparallel is wrapped, with the first forward call it starts splitting into smaller batches to available devices listed in `device_ids`.